File: ua-agent/frontend/src/utility/keyword-crawl/infra/paths.ts
import os from "node:os";
import path from "node:path";

const APP_NAME = "ua-agent";

export function userDataDir(): string {
  if (process.platform === "win32") {
    const appData = process.env["APPDATA"] ?? path.join(os.homedir(), "AppData", "Roaming");
    return path.join(appData, APP_NAME);
  }
  if (process.platform === "darwin") {
    return path.join(os.homedir(), "Library", "Application Support", APP_NAME);
  }
  const xdg = process.env["XDG_DATA_HOME"] ?? path.join(os.homedir(), ".local", "share");
  return path.join(xdg, APP_NAME);
}

export function libraryDbPath(): string {
  return path.join(userDataDir(), "library.db");
}

export function patchrightProfileDir(): string {
  return path.join(userDataDir(), "patchright-profile");
}

export function logsDir(): string {
  return path.join(userDataDir(), "logs");
}

export function keywordCrawlLogPath(): string {
  return path.join(logsDir(), "keyword-crawl.log");
}

export function asrModelDir(): string {
  return path.join(userDataDir(), "asr", "Fun-ASR-Nano-2512");
}

File: ua-agent/backend/src/ua_agent/library/paths.py
from __future__ import annotations

import os
from pathlib import Path

from platformdirs import user_data_dir

_APP_NAME = "ua-agent"


def library_data_dir() -> Path:
    """Return the OS-conventional user-data dir for ua-agent.

    Honors the test override `UA_AGENT_DATA_DIR` so contract tests can pin the
    library DB to a tmp_path without monkeypatching platformdirs.
    """
    override = os.environ.get("UA_AGENT_DATA_DIR")
    base = Path(override) if override else Path(user_data_dir(_APP_NAME, appauthor=False, roaming=False))
    base.mkdir(parents=True, exist_ok=True)
    return base


def library_db_path() -> Path:
    return library_data_dir() / "library.db"


def asr_model_dir() -> Path:
    """Directory holding the Fun-ASR-Nano-2512 model files.

    On Windows we MUST use %APPDATA% (Roaming) to match the Node-side
    `userDataDir()` (frontend/src/utility/keyword-crawl/infra/paths.ts).
    `library_data_dir()` uses %LOCALAPPDATA% via platformdirs default and
    would diverge — Node writes to one location, Python reads from another.
    No bootstrap (does not mkdir) — the downloader owns that.
    """
    if os.name == "nt":
        appdata = os.environ.get("APPDATA")
        if appdata:
            return Path(appdata) / "ua-agent" / "asr" / "Fun-ASR-Nano-2512"
    return library_data_dir() / "asr" / "Fun-ASR-Nano-2512"
